src/build/functions/edge.ts

```diff
@@ -161,26 +194,119 @@ const copyHandlerDependencies = async (
+          parentDirectoriesPath += '../'
+        } else {
+          break
+        }
+      }
+
+      if (file.endsWith('.node')) {
+        // C++ addons are not supported
+        acc.unsupportedDotNodeModules.push(join(srcDir, file))
+      }
+
+      if (dirsUp > acc.maxDirsUp) {
+        return {
+          ...acc,
+          maxDirsUp: dirsUp,
+          maxParentDirectoriesPath: parentDirectoriesPath,
+        }
+      }
+
+      return acc
+    },
+    { maxDirsUp: 0, maxParentDirectoriesPath: '', unsupportedDotNodeModules: [] as string[] },
+  )
+
+  if (unsupportedDotNodeModules.length !== 0) {
+    throw new Error(
+      `Usage of unsupported C++ Addon(s) found in Node.js Middleware:\n${unsupportedDotNodeModules.map((file) => `- ${file}`).join('\n')}\n\nCheck https://docs.netlify.com/build/frameworks/framework-setup-guides/nextjs/overview/#limitations for more information.`,
+    )
+  }
+
+  const commonPrefix = relative(join(srcDir, maxParentDirectoriesPath), srcDir)
+
+  parts.push(`const virtualModules = new Map();`)
+
+  for (const file of files) {
+    const srcPath = join(srcDir, file)
+
+    const content = await readFile(srcPath, 'utf8')
+
+    parts.push(
+      `virtualModules.set(${JSON.stringify(join(commonPrefix, file))}, ${JSON.stringify(content)});`,
+    )
+  }
+  parts.push(`registerCJSModules(import.meta.url, virtualModules);
+
+    const require = createRequire(import.meta.url);
+    const handlerMod = require("./${join(commonPrefix, entry)}");
+    const handler = handlerMod.default || handlerMod;
+
+    export default handler
+    `)
+
+  const outputFile = join(destDir, `server/${name}.js`)
+
+  await mkdir(dirname(outputFile), { recursive: true })
+
+  await writeFile(outputFile, parts.join('\n'))
+}
+
+const createEdgeHandler = async (
+  ctx: PluginContext,
+  definition: EdgeOrNodeMiddlewareDefinition,
+): Promise<void> => {
+  await (definition.runtime === 'edge'
+    ? copyHandlerDependenciesForEdgeMiddleware(ctx, definition.functionDefinition)
+    : copyHandlerDependenciesForNodeMiddleware(ctx))
   await writeHandlerFile(ctx, definition)
 }
 
-const getHandlerName = ({ name }: Pick<NextDefinition, 'name'>): string =>
+const getHandlerName = ({ name }: Pick<EdgeMiddlewareDefinition, 'name'>): string =>
   `${EDGE_HANDLER_NAME}-${name.replace(/\W/g, '-')}`
 
 const buildHandlerDefinition = (
   ctx: PluginContext,
-  { name, matchers, page }: NextDefinition,
+  def: EdgeOrNodeMiddlewareDefinition,
 ): Array<ManifestFunction> => {
-  const functionHandlerName = getHandlerName({ name })
-  const functionName = name.endsWith('middleware')
-    ? 'Next.js Middleware Handler'
-    : `Next.js Edge Handler: ${page}`
-  const cache = name.endsWith('middleware') ? undefined : ('manual' as const)
+  const functionHandlerName = getHandlerName({ name: def.name })
+  const functionName = 'Next.js Middleware Handler'
+  const cache = def.name.endsWith('middleware') ? undefined : ('manual' as const)
```

Comment:
Perhaps we could move this into `EdgeOrNodeMiddlewareDefinition` i.e. `def.cacheName` or similar